### ISSUE TITLE
Getting MIDI note data into FX audio plug in mode.

### ISSUE THREAD

**Issue by pereshaped** [plugdata-team/plugdata#1775]
Im wondering how I might get around this issue? I read the docs and it gives examples how to get MIDI to a Plug Data instrument. Ekkko Station however is an FX AND Instrument. So there are situations where the user can play the delay time of the echo unit using a MIDI keyboard attached to the DAW.

I couldn't, for the life of me, find a way to do this in Logic Pro X, Reaper.

Many thanks

![Screen Shot 2024-07-27 at 12 05 07](https://github.com/user-attachments/assets/d9575bae-918f-47bf-8eb7-7b9e0c189fcb)

**Comment by timothyschoen** [comment]
Hi! I think if you manage to put MIDI into the plugin it should work. Here I'm inputting MIDI into plugdata-fx from within Gig Performer, for example:

https://github.com/user-attachments/assets/fc68dbcc-a217-42af-9db2-b3901668ff13

The main issue here seems to be whether or not the DAW and the plugin format support it.

**Comment by pereshaped** [comment]
MIDI works 100% when I use Plug Data as standalone. As it does in Vanilla.

Note data gets to it when as an AU VST Instr.
It doesn't however, when it as a Host FX(only). I have a huge suite of AUVSTs and all work great.

Tbh, I haven't done extensive tests as Im not a developer and like most musicians, tend to use a prefered DAW, which in this case is Logic Pro X on OsX and AUM, Drambo, Audiobus on iOs but will keep at it.

**Comment by pereshaped** [comment]
So I have now tested the AU plugin version in a number of hosts.

Drambo
Studiomux
AUM
Audiobus.
Logic
Cubasis.

 I can confirm that the **"midiin"** flag is not set to **true** in the AUFX version of the plugin. Furthermore the  iolayout of the MUFX has no inputs they are set to 0. Consequencty: 

1. if you have create an AUFX patch you cannot send MIDI directly to the Pure Data audio fx patch in order to use a hardware MIDI controller. 
2. 
3. If you have created a Pure Data patch which is an instrument but also needs audio in (i.e.) a Granular synth) you cannot send audio to it directly.


**iolayout 0 2;**
code pEre;
type instrument;
**midiin true;**
midiout true;
midionly false;
playhead true;


iolayout 2 2;
code pEre;
type audiofx;
**midiin true;**
midiout true;
midionly false;
playhead true;

**Comment by pereshaped** [comment]
I think I understand this better now.

There is a difference between automation using param objects and using MIDI cc automation.

I can address Plug Data patch with  realtime MIDI CC control or MIDI CCautomation no problem if the instance of Plug Data is loaded as an Instrument.

However, it when loaded as an Audio FX plugin that it would seem that the MidiIn flag is set to "false" in the aufx mode. Thus you cannot send realtime midi for control over the audio fx plugin or note data in all versions of Osx and iOs.

Sorry about all the rambling but it took me a while to understand the difference. I have the issue nailed now

**Comment by timothyschoen** [comment]
> I think I understand this better now.
> 
> There is a difference between automation using param objects and using MIDI cc automation.
> 
> I can address Plug Data patch with realtime MIDI CC control or MIDI CCautomation no problem if the instance of Plug Data is loaded as an Instrument.
> 
> However, it when loaded as an Audio FX plugin that it would seem that the MidiIn flag is set to "false" in the aufx mode. Thus you cannot send realtime midi for control over the audio fx plugin or note data in all versions of Osx and iOs.
> 
> Sorry about all the rambling but it took me a while to understand the difference. I have the issue nailed now

Yes, though more specifically: plugdata will report that it can receive MIDI even if it's an effect plugin. But it depends on your host whether or not it allows this. I think in Logic doesn't but some other DAW/hosts do.

**Comment by pereshaped** [comment]
"... I think in Logic doesn't..." That is completly incorrect Tim. Plug Data is the only plugin when in audio Fx mode that doesn't receive MIDI cc automation data. Every other of the 200 (Waves, Eventide, Lexicon, you name any manufacturer) we have, receive and accept MIDI data and do so fine in Logic (for years) and we use them for production every day. This is a Plug Data issue only. The Plug Data "audiofx" AU header or preferences has "midin" = "false". That is my bet.  And as long as that is the case it will limit the appeal and use of Plug Data as a plugin for audio in DAWs not just Logic. Automation and especially MIDI CC automation is core in DAws. The same lack of MIDI CC input happens in Cubasis (even on Win), Drambo, AUM etc etc etc

**Comment by timothyschoen** [comment]
OK I think I found something, working on it now.

**Comment by pereshaped** [comment]
No worries, Im still testing at this end.

**Comment by timothyschoen** [comment]
I think this should be fixed in the new release!

**Comment by pereshaped** [comment]
Amazing Tim, really! Hopefully it will work on my 10.13 Mac. 0.90 is pretty solid on it.
